controllers/usersController.ts

```diff
@@ -1,73 +1,138 @@
             message:
                 "You can now log in using your registered email or username.",
         });
-    } else {
-        return res.status(500).send({
-            heading: "Server error",
-            message: "Please try again.",
-        });
     }
+
+    return res.status(500).send(SERVER_ERR_OBJ);
 };
 
-export const checkDuplicateEmail: RequestHandler = async (req, res, next) => {
+export const checkDuplicateEmail: RequestHandler = async (
+    req: Request,
+    res: Response,
+    next: NextFunction
+) => {
     const { email, username } = req.body;
 
-    const resultEmail = await prisma.user.findUnique({
+    const resultEmail: { email: string } | null = await prisma.user.findUnique({
         where: { email: email },
         select: {
             email: true,
         },
     });
 
-    const resultUsername = await prisma.user.findUnique({
-        where: { username: username },
-        select: {
-            username: true,
-        },
-    });
+    const resultUsername: { username: string } | null =
+        await prisma.user.findUnique({
+            where: { username: username },
+            select: {
+                username: true,
+            },
+        });
+
+    if (resultEmail) {
+        return res.status(400).send({
+            heading: "Email already taken",
+            message: "Please login or register using another email.",
+        });
+    }
 
-    if (resultEmail || resultUsername) {
+    if (resultUsername) {
         return res.status(400).send({
-            heading: "Duplicate email/username found",
-            message: "Please login.",
+            heading: "Username already taken",
+            message: "Please login or register using another username.",
         });
     }
 
     next();
 };
 
-export const loginUser: RequestHandler = (req, res) => {
-    // Find user, get hashed password and use bcrypt to compare. We use sessions database to store user sesssions?
+export const loginUser: RequestHandler = async (
+    req: Request,
+    res: Response
+) => {
+    const { emailOrUsername, password } = req.body;
+
+    const userFoundByEmail: Prisma.userCreateInput | null =
+        await prisma.user.findUnique({
+            where: { email: emailOrUsername },
+        });
+
+    const userFoundByUsername: Prisma.userCreateInput | null =
+        await prisma.user.findUnique({
+            where: { username: emailOrUsername },
+        });
+
+    let hashFound: string;
+    let user: Prisma.userCreateInput;
+
+    if (userFoundByEmail) {
+        hashFound = userFoundByEmail.password;
+        user = userFoundByEmail;
+    } else if (userFoundByUsername) {
+        hashFound = userFoundByUsername.password;
+        user = userFoundByUsername;
+    } else {
+        return res.status(400).send({
+            heading: "Email or Username is not yet registered",
+            message: "Please register first before logging in.",
+        });
+    }
```

Comment:
```suggestion
    if (userFoundByEmail) {
        hashFound = userFoundByEmail.password;
        user = userFoundByEmail;
    }
    
     if (userFoundByUsername) {
        hashFound = userFoundByUsername.password;
        user = userFoundByUsername;
    }
    
     return { hashFound, user } 
```

this two can be encapsulated in another function

Reply:
What I ended up doing was, which was what I wanted to do initially but couldn't figure out, using the OR option of the where statement. 

This meant that the return value would either be an array with an object inside or an empty array. I decided to chain a then method to return either a null if empty array and the user object if a user was found so that the other codes didn't need to be changed as well as the types.